### PR TITLE
Adapt deployment mode based on the infrastructure topology

### PR DESCRIPTION
**Pull request by simonpasquier**
* [ ] I added CHANGELOG entry for this change.
* [X] No user facing changes, so no entry in CHANGELOG was needed.

This PR implements https://github.com/openshift/enhancements/pull/688 and supersedes #1070.

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -489,7 +489,7 @@ func (o *Operator) Config(key string) (*manifests.Config, error) {
 		klog.Warningf("Could not load proxy configuration from API. This is expected and message can be ignored when proxy configuration doesn't exist. Proceeding without it: %v", err)
 	}
 
-	err = c.LoadPlatform(func() (*configv1.Infrastructure, error) {
+	err = c.LoadInfrastructure(func() (*configv1.Infrastructure, error) {
 		return o.client.GetInfrastructure("cluster")
```
Comment (lilic) [inline pkg/operator/operator.go:493]: Ah nice did not see this one before! 👏

**Review by lilic — COMMENTED** [review]
Great stuff, one comment suggestion otherwise lgtm

File: pkg/manifests/config.go [hunk 2]
```diff
@@ -32,9 +32,13 @@ const (
 )
 
 type Config struct {
-	Images      *Images               `json:"-"`
-	RemoteWrite bool                  `json:"-"`
-	Platform    configv1.PlatformType `json:"-"`
+	Images      *Images `json:"-"`
+	RemoteWrite bool    `json:"-"`
+
+	// Fields retrieved from infrastructure.config.openshift.io/cluster
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/manifests/config.go:38]: agreed, I had the same idea but wanted to test it as-is for now :)

**Comment by simonpasquier** [comment]
@dgrisonnet @paulfantom @sur please have a look. Nothing fancy here but I've refactored a bit to decouple the pkg/client package from pkg/manifests.

You can also have a look at the [enhancement](https://github.com/openshift/enhancements/pull/688).

**Review by dgrisonnet — COMMENTED** [review]
Just a small comment from my side, but it looks good to me.

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -251,14 +263,20 @@ func (o *Operator) keyFunc(obj interface{}) (string, bool) {
 }
 
 func (o *Operator) handleEvent(obj interface{}) {
+	cmoConfigMap := o.namespace + "/" + o.configMapName
+
+	if _, ok := obj.(*configv1.Infrastructure); ok {
+		klog.Infof("Triggering update due to an infrastructure update")
+		o.enqueue(cmoConfigMap)
```
[the remark this answers is not in the document]
Comment (simonpasquier) [reply pkg/operator/operator.go:270]: correct!

**Comment by simonpasquier** [comment]
/hold
until the OEP gets merged

**Comment by simonpasquier** [comment]
/hold cancel
The OEP has been merged and the PR reflects the latest version. Especially CMO remembers the last known infrastructure configuration and if it can't get the current value, it will assume that nothing has changed.

**Comment by dgrisonnet** [comment]
/lgtm
/hold

@openshift/openshift-team-monitoring can we have another pair of :eyes: on this one? Feel free to remove the hold if you think the PR is ready to be merged.

**Pull request merged**